---
title: "[201006] brandi day 23"
categories: etc
---

상품 정보를 insert 하는 DAO 에서 상품 정보를 각각 해당하는 여러 개의 테이블에 입력해주어야 한다.
```py
# 1. products 테이블에 데이터 insert
insert_query = """
INSERT INTO products
(
    created_at,
    is_deleted
) VALUES (
    now(),
    0
)
"""
row = session.execute(insert_query).lastrowid
# 2. product_info 테이블에 데이터 insert
product_info['product_id'] = row
# 이미지 리스트를 가져와 첫 번째 이미지를 대표이미지로 저장
insert_query = """
INSERT INTO product_info
(   
    product_id,
    product_code,
    seller_id,
    is_on_sale,
    is_displayed,
    name,
    simple_description,
    detail_description,
    price,
    discount_rate,
...(생략)
) VALUES (
    :product_id,
    :product_code,
    :seller_id,
    :is_on_sale,
    :is_displayed,
    :name,
    :simple_description,
    :detail_description,
    :price,
...(생략)
)
"""

row = session.execute(insert_query, product_info).lastrowid
```

위와 같이 ```products``` 테이블에서 상품의 id 를 정의해주면 그 아이디를 ```lastrowid``` 로 받아오고 ```product_info``` 테이블의 ```product_id``` 컬럼에 입력해주도록 하였다. 그런데 ```INSERT``` 가 실패한 경우 ```lastlowid``` 가 반환이 되지 않을 수도 있다. 하지만 지금은 ```INSERT``` 마다 이를 처리하는 코드가 없는 상태인데 이에 대한 피드백을 받게 되었다. 

```py
if row <= 0
    raise sqlAlchemyError
```
따라서 위와 같이 쿼리를 실행하고 적용된 row 의 수가 0 이하일 경우에 sqlAlchemy 의 에러를 발생시키려 했는데 그 전에 일부러 에러를 내보고 어떻게 처리되는지 확인을 해보았다. 

이미 있는 row 의 id 에 데이터를 입력하여 일부러 에러를 발생시켰다.

![error](https://user-images.githubusercontent.com/26542094/95579580-e173a700-0a70-11eb-8d82-ca38f267c5f6.png)

포스트맨으로 상품 등록 API 를 호출하고 결과를 확인하면 dao 에서 에러를 잡아주지 않아도 controller 에서 잡은 ```Exeption``` 에서 에러가 잡히는 것을 확인할 수 있었다. 
이 케이스를 통해 확인 할 수 있는 것은 model 또는 service 레이어에서 따로 에러를 잡아주지 않아도 에러가 발생하면 view 레이어로 올라가서 에러를 관리할 수 있다는 것이었다 😮

지금은 저 에러를 따로 잡아주지 않아 sql 문과 함께 에러 메세지가 반환되는데 sql 문이 노출되지 않도록 IntegrityError 등을 세세하게 관리해줄 필요성을 느꼈다!

현재는 아래와 같이 에러 코드를 추가해주었다. 
```py
except exc.IntegrityError:
    return jsonify({'message': 'DUPLICATE_DATA'}), 400
    
except exc.InvalidRequestError:
    return jsonify(({'message' : 'INVALID_REQUEST'})), 400

except exc.ProgrammingError:
    return jsonify(({'message': 'ERROR_IN_SQL_SYNTAX'})), 400
```